### ISSUE TITLE
Cancel all observers

### ISSUE THREAD

**Issue by tbaranes** [postmates/PMKVObserver#14]
It would be great to have a `kvoController.cancelAll()`.

For example, if I have a childViewController which is observing some properties only when it's on screen, then cancelling all each when disappearing, reobserving when reappearing... In the implementation, I have to keep a reference on each token, then cancel them all one by one. It's ok to do this _burden_, but I think that's would be a cool feature for `PMKVObserver` :)

**Comment by lilyball** [comment]
I assume `kvoController.cancelAll()` is a reference to FBKVOController? There's a few things that does that PMKVObserver doesn't, and it would be helpful to know which things you care about:

1. It extends `UIViewController` with a property `kvoController` that basically provides the same thing that PMKVObserver does with the `observer: self` argument.
2. It provides a `cancelAll()`  method that cancels all observers.
3. It de-dupes observer registration, so if you use the same `kvoController` instance and observe the same object/keyPath pair, it only observes once (even if you pass different blocks; only the block passed to the first observation will be invoked).

(1) doesn't seem like it has any benefit over just saying `observer: self` when creating a new `KVObserver`. I suppose it also does let you avoid having to declare `private var observers: [KVObserver] = []` to store your observers, but that hardly seems like a burden.

(2) can be replicated just by keeping an array of observers and then canceling them all. And the array approach also lets you test if you have any current observers by checking `array.isEmpty`. If you don't like having to iterate over the whole array, you could also create your own helper method:

```swift
extension Array where Element: KVObserver {
    mutating func cancelAll() {
        for observer in self {
            observer.cancel()
        }
        removeAll()
    }
}
```

This approach also has the benefit of letting you keep separate arrays of observers that should have distinct lifetimes, whereas with FBKVOController you only get one "free" `kvoController` instance and can't unobserve separate subsets (either unobserve a single object/keyPath, or unobserve everything).

(3) is certainly the hardest to trivially replicate. I don't actually like this behavior because of the fact that it can't take the block into account for de-duping so if you observe with 2 different blocks it just ignores the second (it also ignores the options passed to the second observation). It mostly just seems useful with `UIViewController` where you can do something like set up observations in the view lifecycle methods without having to check if you've already done it (though personally, I don't think having to say `if observers.isEmpty { ... }` is any particular burden).

So which of these behaviors are you actually asking for?

**Comment by tbaranes** [comment]
Thanks for these information! True that I was inspired by `FBKVOController` since I'm using it in another project. Interesting to getting more about the difference between these both libraries.

For this issue, I was mainly looking for the point 1 and 2. Right now, I ended up with the same extension you suggested. Following your comments, I'm not sure if that feature will really fit in `PMKVObserver`, also, as you say, it's not a real _burden_ and allow us to have more control about the observers's lifetime.

Anyway, thanks for that great library!

**Comment by lilyball** [comment]
Thanks for the feedback! Please let me know if you have any other questions or suggestions. I'm going to go ahead and close this ticket, as I don't think PMKVObserver should be providing that `Array` extension itself, nor do I really want to introduce a `KVObserverGroup` class that acts like an `Array` just to have `cancelAll()` method. If someone has a genuine use for de-duping observers, then I'll consider adding a helper class for dealing with that, but not until I understand why someone wants it, and that can be dealt with in a separate ticket (if it happens at all).
